src/node/consumer/src/lib/services/metadata_service.ts

```diff
@@ -0,0 +1,234 @@
+import {TorrentType} from '@enums/torrent_types';
+import {ICacheService} from "@interfaces/cache_service";
+import {ICinemetaJsonResponse} from "@interfaces/cinemeta_metadata";
+import {ICommonVideoMetadata} from "@interfaces/common_video_metadata";
+import {IKitsuCatalogJsonResponse} from "@interfaces/kitsu_catalog_metadata";
+import {IKitsuJsonResponse} from "@interfaces/kitsu_metadata";
+import {IMetaDataQuery} from "@interfaces/metadata_query";
+import {IMetadataResponse} from "@interfaces/metadata_response";
+import {IMetadataService} from "@interfaces/metadata_service";
+import {IocTypes} from "@models/ioc_types";
+import axios from 'axios';
+import {ResultTypes, search} from 'google-sr';
+import {inject, injectable} from "inversify";
+import nameToImdb from 'name-to-imdb';
+
+const CINEMETA_URL = 'https://v3-cinemeta.strem.io';
+const KITSU_URL = 'https://anime-kitsu.strem.fun';
+const TIMEOUT = 60000;
+
+@injectable()
+export class MetadataService implements IMetadataService {
+    private cacheService: ICacheService;
+
+    constructor(@inject(IocTypes.ICacheService) cacheService: ICacheService) {
+        this.cacheService = cacheService;
+    }
+
+    public getKitsuId = async (info: IMetaDataQuery): Promise<number | Error> => {
```

Comment:
The default access modifier is `public` so you can omit that. And you don't need `=` to define a class method. It probably works because JS is weird but best stick to TS conventions.

```suggestion
    async getKitsuId(info: IMetaDataQuery): Promise<number | Error> => {
```

Reply:
I've done this for all public methods

All private methods i've had to leave as arrow functions, in order to avoid the context binding requirements we'd have to setup to call them with `this.`

Any of the injected services would be undefined without setting up bindings like this.someMethod.bind(this); for each method in the constructor

Reply:
Hmm interesting. I'd like to explore this locally tomorrow, see if it can still be improved somehow. Not that it's bad per se, just looks a bit weird this way.

Reply:
came from this: https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript
When i had failing tests

Easiest way to test is the tracker service tests, as there is one public method and one private method, if you change the private method from an arrow function to a class method, the logger will be undefined unless you setup the binding context for it in the constructor

Reply:
```js
import {ICacheService} from "@interfaces/cache_service";
import {ILoggingService} from "@interfaces/logging_service";
import {ITrackerService} from "@interfaces/tracker_service";
import {configurationService} from '@services/configuration_service';
import {IocTypes} from "@setup/ioc_types";
import axios, {AxiosResponse} from 'axios';
import {inject, injectable} from "inversify";

@injectable()
export class TrackerService implements ITrackerService {
    @inject(IocTypes.ICacheService) cacheService: ICacheService;
    @inject(IocTypes.ILoggingService) logger: ILoggingService;

    async getTrackers(): Promise<string[]> {
        return this.cacheService.cacheTrackers(this.downloadTrackers);
    }

    private async downloadTrackers(): Promise<string[]> {
        const response: AxiosResponse<string> = await axios.get(configurationService.trackerConfig.TRACKERS_URL);
        const trackersListText: string = response.data;
        // Trackers are separated by a newline character
        let urlTrackers = trackersListText.split("\n");
        // remove blank lines
        urlTrackers = urlTrackers.filter(line => line.trim() !== '');

        if (!configurationService.trackerConfig.UDP_ENABLED) {
            // remove any udp trackers
            urlTrackers = urlTrackers.filter(line => !line.startsWith('udp://'));

        }

        this.logger.info(`Trackers updated at ${Date.now()}: ${urlTrackers.length} trackers`);

        return urlTrackers;
    }
}
```

as opposed to

```js
import {ICacheService} from "@interfaces/cache_service";
import {ILoggingService} from "@interfaces/logging_service";
import {ITrackerService} from "@interfaces/tracker_service";
import {configurationService} from '@services/configuration_service';
import {IocTypes} from "@setup/ioc_types";
import axios, {AxiosResponse} from 'axios';
import {inject, injectable} from "inversify";

@injectable()
export class TrackerService implements ITrackerService {
    @inject(IocTypes.ICacheService) cacheService: ICacheService;
    @inject(IocTypes.ILoggingService) logger: ILoggingService;

    async getTrackers(): Promise<string[]> {
        return this.cacheService.cacheTrackers(this.downloadTrackers);
    }

    private downloadTrackers = async (): Promise<string[]> => {
        const response: AxiosResponse<string> = await axios.get(configurationService.trackerConfig.TRACKERS_URL);
        const trackersListText: string = response.data;
        // Trackers are separated by a newline character
        let urlTrackers = trackersListText.split("\n");
        // remove blank lines
        urlTrackers = urlTrackers.filter(line => line.trim() !== '');

        if (!configurationService.trackerConfig.UDP_ENABLED) {
            // remove any udp trackers
            urlTrackers = urlTrackers.filter(line => !line.startsWith('udp://'));

        }

        this.logger.info(`Trackers updated at ${Date.now()}: ${urlTrackers.length} trackers`);

        return urlTrackers;
    };
}
```

Reply:
Binding looks cleaner imo, it's a native js feature whereas the other approach is kind of a hack to avoid using bind.

Reply:
Not on pc rn so can't test but does `this.cacheService.cacheTrackers((...) => this.downloadTrackers(...)` work?

Reply:
It does but the problem with binding is there is no type safety inside the parameters passed to the function, not on the function itself so you can pass anything you want

An arrow isn't really a hack it's just a function with global context ^^ and is native too. 

I agree it would be nicer to use bind, but only if the caller also could adhere to type safety and not just once you hit the called method

Reply:
Using the arrow just preserves the lexical this of the owning class

Of we want type safety and want to bind then we have to bind the context in the method and imo that looks a bit messy

Like this:

'''
class Dog {
  public constructor (private name: string) { }
  private eat (this: Dog, food: string): void { // annotate this and food parameters
    console.log (`${this.name} is eating ${food}.`);
  }
  public run (): void {
    this.eat ('bone'); // OK
  }
}
'''

Reply:
I was confused because the function looks like a class method but due to the assignment it's actually more like a property assignment, which is still valid JavaScript. If you separate this snippet and have a look at the transpiled code then the following typescript:

```ts
class Foo {
    private bar = (): number => {
        return 0;
    }

    private baz(): number {
        return 1;
    }
}
```

Becomes: 

```js
var Foo = /** @class */ (function () {
    function Foo() {
        this.bar = function () {
            return 0;
        };
    }
    Foo.prototype.baz = function () {
        return 1;
    };
    return Foo;
}());
```

Meh, it's probably fine to keep it the way you originally did then, only looks a bit weird in my book but ultimately it does what it's supposed to do.

Reply:
Cool cheers :)